### PR TITLE
feat: Add /api/v2/secret endpoint to store a secret

### PR DESCRIPTION
**Pull request by lenny-intel**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/device-sdk-c/blob/master/.github/CONTRIBUTING.md

## What is the current behavior?
No way to store secrets for the instance of a device service 

## What is the new behavior?

/api/v2/secret endpoint add to store exclusive secrets for the running device service instance.

This PR is dependent on PR https://github.com/edgexfoundry/device-sdk-go/pull/707

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## New Imports

- [ ] Yes
- [x] No

### CODE REVIEW

**Review by hahattan — DISMISSED** [review]
LGTM

**Comment by cloudxxx8** [comment]
@lenny-intel please help rebase and resolve the conflicts.  thanks.

File: example/cmd/device-simple/Attribution.txt [hunk 1]
```diff
@@ -130,4 +130,7 @@ go-playground/validator/v10 (MIT) https://github.com/go-playground/validator/v10
 https://github.com/go-playground/validator/blob/master/LICENSE
 
 leodido/go-urn (MIT) https://github.com/leodido/go-urn
-https://github.com/leodido/go-urn/blob/master/LICENSE
\ No newline at end of file
+https://github.com/leodido/go-urn/blob/master/LICENSE
+
+stretchr/objx (MIT) https://github.com/stretchr/objx
+https://github.com/stretchr/objx/blob/master/LICENSE
```
Comment (lenny-intel) [inline example/cmd/device-simple/Attribution.txt:136]: Not sure why this just now was identified in this PR, but it is required for PR to pass verify checks.

File: openapi/v2/device-sdk.yaml [hunk 2]
```diff
@@ -481,6 +481,39 @@ components:
...
               description: "A string representation of the reading's value"
               type: string
+    SecretsRequest:
+      allOf:
+        - $ref: '#/components/schemas/BaseRequest'
+      description: Defines the secret data to be stored
+      type: object
+      properties:
+        path:
+          description: Specifies the type or location in the secret to store
```
Comment (tonyespy) [inline openapi/v2/device-sdk.yaml:491]: type or location? And also shouldn't be "...of the secret(s) to store"?

**Review by tonyespy — COMMENTED** [review]
A couple of minor comments inline...  Testing works as expected, although I was surprised to see the apiVersion returned as "v2" vs. "2".

File: internal/v2/controller/http/common_test.go [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+
+	_, err = time.Parse(time.UnixDate, actual.Timestamp)
+	assert.NoError(t, err)
+
+	require.Equal(t, contractsV2.ApiVersion, actual.ApiVersion)
+}
+
+func TestVersionRequest(t *testing.T) {
+	expectedServiceVersion := "1.2.5"
+	expectedSdkVersion := "1.3.1"
```
[the remark this answers is not in the document]
Comment (lenny-intel) [reply internal/v2/controller/http/common_test.go:74]: `expectedServiceVersion` is the version of the device service. This is a unit test so there isn't an actual device service and these are not actual version numbers, i.e. not impacted the real SDK verion.

File: openapi/v2/device-sdk.yaml (hunk 2 above)
Comment (lenny-intel) [reply openapi/v2/device-sdk.yaml:491]: Yea, should just be location. Will reword. Remember the ADR definition of a secret? ;-) a collection of key value pairs.

Comment (lenny-intel) [reply openapi/v2/device-sdk.yaml:491]: And per that definition, the endpoint should be `/secret`, not `/secrets`. Reworking and will submit issue to do same for App services.

File: internal/v2/controller/http/common.go [hunk 4]
```diff
@@ -55,3 +60,51 @@ func (c *V2HttpController) Metrics(writer http.ResponseWriter, request *http.Req
...
+	provider := bootstrapContainer.SecretProviderFrom(c.dic.Get)
+	secretRequest := common.SecretRequest{}
+	err := json.NewDecoder(request.Body).Decode(&secretRequest)
+	if err != nil {
+		edgexError := errors.NewCommonEdgeX(errors.KindContractInvalid, "JSON decode failed", err)
+		c.sendEdgexError(writer, request, edgexError, sdkCommon.APIV2SecretRoute)
+		return
+	}
+
+	path, secrets := c.prepareSecret(secretRequest)
```
Comment (tonyespy) [inline internal/v2/controller/http/common.go:80]: nit -> secret?

File: internal/v2/controller/http/common.go [hunk 5]
```diff
@@ -55,3 +60,51 @@ func (c *V2HttpController) Metrics(writer http.ResponseWriter, request *http.Req
...
+
+	if err := provider.StoreSecrets(path, secrets); err != nil {
```
Comment (tonyespy) [inline internal/v2/controller/http/common.go:82]: same as above

File: internal/v2/controller/http/common.go [hunk 6]
```diff
@@ -55,3 +60,51 @@ func (c *V2HttpController) Metrics(writer http.ResponseWriter, request *http.Req
...
+		edgexError := errors.NewCommonEdgeX(errors.KindServerError, "Storing secrets failed", err)
```
Comment (tonyespy) [inline internal/v2/controller/http/common.go:83]: same as above

File: internal/v2/controller/http/common_test.go [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+	require.Less(t, 0, len(configJson))
+
+	actualConfig := sdkCommon.ConfigurationStruct{}
+	err = json.Unmarshal(configJson, &actualConfig)
+	require.NoError(t, err)
+
+	assert.Equal(t, expectedConfig, actualConfig)
+}
+
+func TestSecretsRequest(t *testing.T) {
```
Comment (tonyespy) [inline internal/v2/controller/http/common_test.go:166]: Test**Secret**Request?

File: internal/v2/controller/http/common_test.go [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+
+	NoPath := validRequest
+	NoPath.Path = ""
+	validPathWithSlash := validRequest
+	validPathWithSlash.Path = "/mqtt"
+	validNoRequestId := validRequest
+	validNoRequestId.RequestId = ""
+	badRequestId := validRequest
+	badRequestId.RequestId = "bad requestId"
+	noSecrets := validRequest
```
Comment (tonyespy) [inline internal/v2/controller/http/common_test.go:206]: **noSecret**?

File: internal/v2/controller/http/common_test.go [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+		{Key: "username", Value: ""},
+	}
+	noSecretStore := validRequest
+	noSecretStore.Path = "no"
+
+	tests := []struct {
+		Name               string
+		Request            common.SecretRequest
+		ExpectedRequestId  string
+		SecretsPath        string
```
Comment (tonyespy) [inline internal/v2/controller/http/common_test.go:223]: SecretPath?

File: internal/v2/controller/http/common_test.go [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+		SecretStoreEnabled string
+		ErrorExpected      bool
+		ExpectedStatusCode int
+	}{
+		{"Valid - sub-path no trailing slash, SecretsPath has trailing slash", validRequest, expectedRequestId, "my-secrets/", "true", false, http.StatusCreated},
```
Comment (tonyespy) [inline internal/v2/controller/http/common_test.go:228]: same as above

**Review by tonyespy — COMMENTED** [review]
A few minor comments inline...

File: internal/v2/controller/http/common.go (hunk 5 above)
Comment (lenny-intel) [reply internal/v2/controller/http/common.go:82]: Hehe, I got the function name but missed the var name....

File: internal/v2/controller/http/common.go (hunk 4 above)
Comment (lenny-intel) [reply internal/v2/controller/http/common.go:80]: Hehe, I got the function name but missed the var name....

**Review by tonyespy — APPROVED** [review]
LGTM

**Pull request merged**